### ISSUE TITLE
Documentation for `await_only_futures` is lacking.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/linter#2644]
The documentation for `await_only_futures` doesn't address any of a number of relevant edge cases.

The example given is pretty useless. The fix for `await 23` is to write `23`, not `await Future.value(23)`.
You should never write `await Future.value(x)` in real code either, though it might make sense in tests.
If you want a microtask delay inserted, don't mix that with a value computation, just do `await Future.microtask(() {});` as a statement.

It's unclear how this lint interacts with types like `FutureOr<X>` (it should be fine to await that) or `Future<X>?` (probably fine to await that too) or even `await dynamicExpression` (seem reasonable since it could be a `Future`). 
It actually accepts all of these, so that's good, but not clear from the documentaiton.

I'm on the fence about `await neverExpression`, but since `Never` *is* a subtype of `Future`, and it probably should be accepted like we accept `Never` in any other situation. It currently isnt.

So, proposal:

>  **AVOID** using await on anything which is not a future.
>
> Types that are counts as (likely) being futures are 
> `Future<X>`, `FutureOr<X>`, `Future<X>?`, `FutureOr<X>?` and `dynamic`.
>
> Further, using `await null` is specifically allowed as a way to introduce a microtask delay.
> 
> **BAD:**
> ```dart
> main() async {
>   print(await 23);
> }
> ```
> **GOOD:**
> ```dart
>  main() async {
>    await null; // If a delay is really intended.
>    print(23);
> }
> ```

**Comment by pq** [comment]
This is fantastic.  Thank you @lrhn!  I expect these lacking docs have been perpetuating confusion for a lot of folks.

/fyi @davidmorgan

**Comment by davidmorgan** [comment]
LG, thanks! Except for a typo: "Types that are counts as" ... maybe simplify to something like "Await is allowed on the types:"

**Comment by lrhn** [comment]
Feel free to improve the proposal! 😁
